llvm/test/Analysis/CostModel/RISCV/shuffle-exact-vlen.ll

```diff
@@ -737,8 +737,8 @@ define void @multipart() vscale_range(2,2) {
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %v16b = shufflevector <8 x i16> poison, <8 x i16> poison, <8 x i32> <i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1>
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 12 for instruction: %v16c = shufflevector <16 x i16> poison, <16 x i16> poison, <16 x i32> <i32 0, i32 8, i32 1, i32 9, i32 2, i32 10, i32 3, i32 11, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1>
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 12 for instruction: %v16d = shufflevector <16 x i16> poison, <16 x i16> poison, <16 x i32> <i32 0, i32 16, i32 1, i32 17, i32 2, i32 18, i32 3, i32 19, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1>
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %v32a = shufflevector <4 x i32> poison, <4 x i32> poison, <4 x i32> <i32 2, i32 3, i32 0, i32 1>
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 16 for instruction: %v32a4 = shufflevector <16 x i32> poison, <16 x i32> poison, <16 x i32> <i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1>
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 5 for instruction: %v32a = shufflevector <4 x i32> poison, <4 x i32> poison, <4 x i32> <i32 2, i32 3, i32 0, i32 1>
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 20 for instruction: %v32a4 = shufflevector <16 x i32> poison, <16 x i32> poison, <16 x i32> <i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1>
```

Comment:
This shuffle is a gather, how come it ends up getting changed here? Is this somehow passing through isMaskedSlidePair?

Reply:
Yeah it's not a vrgather.vi but a vrgatherei16.vv at e64, I'm just confused as to how the changes in this PR end up affecting its cost. 

The only change in getShuffleCost is that we now detect isMaskedSlidePair, but I don't see how this would be categorised a pair of slides, even an element rotate, given that it has lanes repeating more than twice?

Reply:
The actual lowering for this is:
```
define <16 x i32> @luke2(<16 x i32> %a) vscale_range(2,2) {
; CHECK-LABEL: luke2:
; CHECK:       # %bb.0:
; CHECK-NEXT:    vsetivli zero, 4, e32, m1, ta, ma
; CHECK-NEXT:    vslidedown.vi v12, v8, 2
; CHECK-NEXT:    vslideup.vi v12, v8, 2
; CHECK-NEXT:    vmv.v.v v13, v12
; CHECK-NEXT:    vmv.v.v v14, v12
; CHECK-NEXT:    vmv.v.v v15, v12
; CHECK-NEXT:    vmv4r.v v8, v12
; CHECK-NEXT:    ret
  %res = shufflevector <16 x i32> %a, <16 x i32> poison, <16 x i32> <i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1>
  ret <16 x i32> %res
}
```
Note this is a VLS shuffle, and it's being split at register boundaries.  The costing parallels that via costShuffleViaVRegSplitting

Reply:
A further bit of context because this confused me at first.

The VLS splitting code is not taking advantage of reuse in the destination result, and is instead costing the shuffle "2 3 0 1" 4 separate times in this example.  We should fix that (separately).

With this change, the cost of that shuffle is 5.  Before, the cost was 4.  As far as the existing instruction costing is concerned, vrgather.vv and vslideup/down have the same cost.  We should fix that (separately).

Reply:
Ah I missed the vscale_range(2,2), that explains it. +1 on fixing separately.